pgjdbc/src/test/java/org/postgresql/test/jdbc2/TimezoneCachingTest.java

```diff
@@ -0,0 +1,411 @@
+      pstmt.setInt(1, 1);
+      // We are in tz1, so timestamp added as tz1.
+      pstmt.setTimestamp(2, ts1);
+      pstmt.addBatch();
+      pstmt.executeBatch();
+      stmt = con.createStatement();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertEquals(
+          "Current TZ is tz1, empty cache to be initialized to tz1 => retrieve in tz1, timestamps must be equal",
+          ts1, rs.getTimestamp(2));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      TimeZone.setDefault(tz2);
+      assertEquals(
+          "Current TZ is tz2, empty cache to be initialized to tz2 => retrieve in tz2, timestamps cannot be equal",
+          ts2, rs.getTimestamp(2));
+      assertEquals(
+          "Explicit tz1 calendar, so timestamps must be equal",
+          ts1, rs.getTimestamp(2, c1));
+      assertEquals(
+          "Cache was initialized to tz2, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2));
+      TimeZone.setDefault(tz1);
+      assertEquals(
+          "Cache was initialized to tz2, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2, c2));
+      assertEquals(
+          "Current TZ is tz1, empty cache to be initialized to tz1 => retrieve in tz1, timestamps must be equal",
+          ts1, rs.getTimestamp(2));
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          ts2, rs.getTimestamp(2, c2));
+      assertEquals(
+          "Explicit tz2 calendar, so timestamps must be equal",
+          ts1, rs.getTimestamp(2, c1));
+      rs.close();
+    } finally {
+      TimeZone.setDefault(null);
+      TestUtil.closeQuietly(rs);
+      TestUtil.closeQuietly(pstmt);
+      TestUtil.closeQuietly(stmt);
+    }
+  }
+
+  private void checkTimestamp(String checkText, Statement stmt, Timestamp ts, TimeZone tz)
+      throws SQLException {
+    TimeZone prevTz = TimeZone.getDefault();
+    TimeZone.setDefault(tz);
+    ResultSet rs = stmt.executeQuery("SELECT col2 FROM testtz");
+    rs.next();
+    Timestamp dbTs = rs.getTimestamp(1);
+    rs.close();
+    TimeZone.setDefault(prevTz);
+    assertEquals(checkText, ts, dbTs);
+  }
+
+  private TimeZone getTimeZoneCache(Object stmt) {
+    try {
+      Field defaultTimeZoneField = stmt.getClass().getDeclaredField("defaultTimeZone");
+      defaultTimeZoneField.setAccessible(true);
+      return (TimeZone) defaultTimeZoneField.get(stmt);
+    } catch (Exception e) {
+    }
+    return null;
+  }
+
+  public TimezoneCachingTest(String name) {
+    super(name);
+    try {
+      Class.forName("org.postgresql.Driver");
```

Comment:
It comes from original `BatchedInsertReWriteEnabledTest` which was in master. I cleaned up `BatchedInsertReWriteEnabledTest` when I realized it was not needed but forgot that one. Will do.